### PR TITLE
feat(mknod): add test from mknod/11.t

### PR DESCRIPTION
**Pull request by musikid**
Also add makedev, major and minor for FreeBSD and Solaris (untested) platfoms.

### CODE REVIEW

File: rust/src/utils/dev.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    pub fn major(dev: nix::libc::dev_t) -> u64 {
+        nix::sys::stat::major(dev)
+    }
+
+    pub fn minor(dev: nix::libc::dev_t) -> u64 {
+        nix::sys::stat::minor(dev)
+    }
+}
+
+// TODO: Upstream to nix/libc?
```
Comment (asomers) [inline rust/src/utils/dev.rs:25]: Yes.  I'll work on this.

File: rust/src/utils/dev.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+// TODO: Would be clearer with cfg_if
+
+#[cfg(target_os = "freebsd")]
+pub use freebsd::*;
+#[cfg(target_os = "linux")]
+pub use linux::*;
+#[cfg(target_os = "solaris")]
```
Comment (asomers) [inline rust/src/utils/dev.rs:7]: Why bother with Solaris?  If anything, you should use `illumos`.  In Rust, those are considered separate operating systems.

File: rust/src/tests/mknod.rs [hunk 3]
```diff
@@ -52,3 +61,106 @@ fn changed_time_fields_success(ctx: &mut TestContext) {
...
+    )
+    .is_ok());
+
+    let stat = symlink_metadata(&file).unwrap();
+    assert_eq!(stat.permissions().mode() as mode_t & ALLPERMS, mode);
+    assert_eq!(major(stat.rdev()) as u64, major_num as u64);
+    assert_eq!(minor(stat.rdev()) as u64, minor_num as u64);
+    assert!(check(&stat.file_type()));
+
+    assert_eq!(
```
Comment (asomers) [inline rust/src/tests/mknod.rs:102]: The EEXIST stuff is totally redundant with what's in 08.t.  I suggest deleting it here.  If you do keep it, you should move it to a separate test function.

File: rust/src/tests/mknod.rs [hunk 4]
```diff
@@ -52,3 +61,106 @@ fn changed_time_fields_success(ctx: &mut TestContext) {
...
+            Mode::from_bits_truncate(mode),
+            makedev(3, 4)
+        ),
+        Err(Errno::EEXIST)
+    );
+
+    // TODO: EINVAL seems to be sent by makedev?
+    #[cfg(target_os = "solaris")]
+    {
+        let file = ctx.gen_path();
```
Comment (asomers) [inline rust/src/tests/mknod.rs:115]: This is functionally independent, so it should be its own test case.

File: rust/src/tests/mknod.rs [hunk 5]
```diff
@@ -52,3 +61,106 @@ fn changed_time_fields_success(ctx: &mut TestContext) {
...
+                argument,
+                Mode::from_bits_truncate(mode),
+                makedev(4000, 4000)
+            ),
+            Err(Errno::EEXIST)
+        );
+
+        let file = ctx.gen_path();
+
+        assert_eq!(
```
Comment (asomers) [inline rust/src/tests/mknod.rs:142]: This part too should have its own test case

File: rust/src/tests/mknod.rs [hunk 6]
```diff
@@ -52,3 +61,106 @@ fn changed_time_fields_success(ctx: &mut TestContext) {
...
+                &file,
+                argument,
+                Mode::from_bits_truncate(mode),
+                makedev(4096, 262144)
+            ),
+            Err(Errno::EINVAL)
+        );
+    }
+
+    let file = ctx.gen_path();
```
Comment (asomers) [inline rust/src/tests/mknod.rs:153]: Finally, this part should get its own test case too.

**Review by asomers — CHANGES_REQUESTED** [review]

File: rust/src/utils/dev.rs [hunk 7]
```diff
@@ -22,8 +22,6 @@ mod linux {
     }
 }
 
-// TODO: Upstream to nix/libc?
-
```
Comment (asomers) [inline rust/src/utils/dev.rs]: This is in progress at https://github.com/rust-lang/libc/pull/2891 .

**Review by asomers — COMMENTED** [review]
This LGTM now, except for the redefinition of makedev.  Let's wait until that libc PR merges.

**Comment by asomers** [comment]
Ok, the `makedev` changes are in libc.  You should be able to update now.

File: rust/Cargo.toml [hunk 8]
```diff
@@ -19,6 +18,11 @@ serde = { version = "1.0.138", features = ["derive"] }
 inventory = "0.3.0"
 walkdir = "2.3.2"
 
+[dependencies.libc]
+git = "https://github.com/rust-lang/libc"
+# TODO: Set version with new makedev platforms
```
Comment (asomers) [inline rust/Cargo.toml:23]: What does this TODO mean?

Comment (musikid) [reply rust/Cargo.toml:23]: Just to change the version when a new libc version will be released with the makedev functions.

**Review by asomers — APPROVED** [review]

**Pull request merged**